### ISSUE TITLE
avoid calling Sys.getlocale("LC_MESSAGES") on Windows

### ISSUE THREAD

**Issue by kevinushey** [s-u/rJava#296]
We recently received a bug report on the RStudio IDE side (https://github.com/rstudio/rstudio/issues/11076), indicating a crash when calling `rJavaj::.jinit()` from within RStudio with R 4.2.0 on Windows.

The crash ultimately appears to be coming from a call to `Sys.getlocale("LC_MESSAGES")`, which is happening here:

https://github.com/s-u/rJava/blob/master/R/jinit.R#L51-L55

Given that LC_MESSAGES (+ LC_PAPER, LC_MEASUREMENT) are not supported on Windows, would you welcome a PR removing those categories on Windows?

For posterity, the crash doesn't appear to happen with R 4.2.0 with Rterm nor Rgui, but that may just be by chance.

**Comment by s-u** [comment]
Well, if it is in `Sys.getlocale()` then it has nothing to with `rJava` - either an RStudio bug or R since `Sys.getlocale()` certainly does have crash as defined behavior. Also if it is rJava related, please file actual facts (like the actual crash report, how to reproduce, `sessionInfo()` etc.). I'll close this for now - please feel free to reopen if more actionable information is provided.

**Comment by kevinushey** [comment]
Thanks. The issue is now resolved in R-devel via https://github.com/wch/r-source/commit/1b5c77482d441074ea4c7624665908e3b671b0c2.

If you receive any bug reports related to rJava + R 4.2.0 + RStudio, you should hopefully be able to instruct them to use R 4.2.0-patched.

**Comment by s-u** [comment]
Did you even find out why it crashed RStudio and nothing else? I suspect something more sinister there since this may be just the tip of the iceberg with other crashes in RStudio lurking. It would be likely better to find the problem...

**Comment by kevinushey** [comment]
We did -- the documentation in https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/setlocale-wsetlocale?view=msvc-170 has:

> This function validates the category parameter. If the category parameter isn't one of the values given in the previous table, the invalid parameter handler is invoked, as described in [Parameter Validation](https://docs.microsoft.com/en-us/cpp/c-runtime-library/parameter-validation?view=msvc-170). If execution is allowed to continue, the function sets errno to EINVAL and returns NULL.

Note that, depending on your interpretation, this would contradict the documented return value earlier in the page:

> If a valid locale and category are given, returns a pointer to the string associated with the specified locale and category. If the locale or category isn't valid, returns a null pointer, and the current locale settings of the program are unchanged.

And the default "invalid parameter handler" is described as:

> The invalid parameter handler dispatch function calls the currently assigned invalid parameter handler. By default, the invalid parameter calls _invoke_watson, which causes the application to close and generate a mini-dump. If enabled by the operating system, a dialog box asks the user if they want to send the crash dump to Microsoft for analysis.

However, it seems like MinGW sets its own no-op invalid parameter handler by default; e.g.

https://github.com/mirror/mingw-w64/blob/f66b1857778bbfc3ab456ff3771f918d122cb475/mingw-w64-crt/crt/crtexe.c#L112-L123

https://github.com/mirror/mingw-w64/blob/f66b1857778bbfc3ab456ff3771f918d122cb475/mingw-w64-crt/crt/crtexe.c#L274

but that's not set for RStudio's `rsession` builds, presumedly because we build using MSVC rather than Rtools.

So, in short -- the windows runtime checks for invalid categories in `setlocale()`, and calls an (overridable) invalid_parameter hook whose default behavior closes the application and generates a minidump. This behavior is not observed in R, presumedly because MinGW overrides that hook with its own no-op handler.

**Comment by s-u** [comment]
OK, thanks, in the meantime I had a discussion offline with Tomas who provided more details and an actual list of the locale settings (not)supported so there should be a work-around shortly. So I guess it would be good to fix RStudio by not segfaulting in the error handlers as that may happen in other Windows APIs. In the meantime we have a work-around in R and then here, so many band-aids ;).
